File: leetcode-submissions/1061_Lexicographically_Smallest_Equivalent_String.ts
function smallestEquivalentString(s1: string, s2: string, baseStr: string): string {
    const adj:Map<string , string[]>= new Map();
    const n:number = s1.length;
    for(let i:number =0 ;i<n;i++){
        if(!adj.has(s1[i])){
            adj.set(s1[i], []);

        }
        if(!adj.has(s2[i])){
            adj.set(s2[i] ,[]);
        }
       adj.get(s1[i])!.push(s2[i]);
        adj.get(s2[i])!.push(s1[i]);

    }
    let ans:string = "";
    for(let i:number =0;i<baseStr.length;i++ ){
        const visited:boolean[]= new Array(26).fill(false);
        const ch :string = baseStr[i];
         if (!adj.has(ch)) {
            ans += ch;
        } else {
            ans += dfs(adj, visited, ch);
        }
    }
    return ans ;
};
function dfs(adj:Map<string , string[]>, visited:boolean[], curr:string):string {
    const index = curr.charCodeAt(0)-"a".charCodeAt(0);
    visited[index]= true ;
    let minCharacter :string = curr; 
    //ab isko mark karne ke baad neighbors explore karlo 
    for(const nbr of adj.get(curr)!){
        const Nbrindex = nbr.charCodeAt(0)-"a".charCodeAt(0);
        if(visited[Nbrindex]==false){
             const ans :string = dfs(adj , visited , nbr);
            if(minCharacter>ans){
                minCharacter = ans ;
            }
            
        }
    }
    return minCharacter;
}

File: leetcode-submissions/1061_Lexicographically_Smallest_Equivalent_String.cpp
class Solution {
public:
    char DFS(unordered_map<char, vector<char>>&adj , char curr, vector<bool>&visited){
        visited[curr-'a']= true;
        char minChar = curr;
        for(char& nbr : adj[curr]){
            if(not visited[nbr-'a']){
                minChar = min(minChar, DFS(adj , nbr, visited));

            }
        }
        return minChar ;

    }
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        unordered_map<char , vector<char>>adj ;
        int n = s1.size();
        int m = baseStr.size();
        for(int i=0;i<n;i++){
            char u = s1[i];
            char v = s2[i];
            adj[u].push_back(v);
            adj[v].push_back(u);

        }
        
        for(int i=0;i<m ;i++){
            vector<bool>visited(26, false);
            baseStr[i]= DFS(adj , baseStr[i],visited);

        }
        return baseStr;

    }
};